### PR TITLE
Add recursive on version bump

### PR DESCRIPTION
**Pull request by tbuatois**
Automatic detection of CID major version bump and ask user for recursive update. 

*Description of changes:*
- Various bug fixes 
- Addition of cid_version on template class from Description parameters 
- Additional information about cid_version in status command 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cid/helpers/quicksight/template.py [hunk 1]
```diff
@@ -1,16 +1,82 @@
...
+    def get_version_as_tuple(self) -> tuple:
+        return (self.major,self.minor,self.build)
 class Template(CidQsResource):
 
     @property
     def id(self) -> str:
         return self.get_property('TemplateId')
+    
+    @property
+    def arn(self) -> str:
```
Comment (darken99) [inline cid/helpers/quicksight/template.py:78]: https://github.com/aws-samples/aws-cudos-framework-deployment/blob/9dac00ac6543579552880bcbe09351a8ed61cd4a/cid/helpers/quicksight/resource.py#L12

File: cid/export.py [hunk 2]
```diff
@@ -146,7 +148,7 @@ def export_analysis(qs):
         TemplateId=template_id,
         GrantPermissions=[
             {
-                "Principal": {"AWS": reader_account} if reader_account != '*' else '*',
+                "Principal": f'arn:aws:iam::{reader_account_id}:root' if reader_account_id != '*' else '*',
```
Comment (iakov-aws) [inline cid/export.py:151]: Good catch

File: cid/export.py [hunk 3]
```diff
@@ -39,11 +39,13 @@ def export_analysis(qs):
...
                 analyzes.extend(page.get('AnalysisSummaryList'))
             if len(analyzes) == 100:
-                logger.info('Too many analyzes. Will show first 100')
+                logger.info('Too many analyses. Will show first 100')
         except qs.client.exceptions.AccessDeniedException:
             logger.info("AccessDeniedException while discovering analyses")
         else:
             analyzes = list(filter(lambda a: a['Status']=='CREATION_SUCCESSFUL', analyzes ))
+            if not analyzes:
+                raise CidCritical("No analyses was found, please save your dashboard as an analyse first")
```
Comment (iakov-aws) [inline cid/export.py:48]: Agree

File: cid/helpers/quicksight/template.py [hunk 4]
```diff
@@ -1,16 +1,82 @@
 import logging
+import re
 from typing import Dict
-
 from cid.helpers.quicksight.resource import CidQsResource
 
 logger = logging.getLogger(__name__)
 
+class CidVersion:
+    def __init__(self, str_version):
```
Comment (iakov-aws) [inline cid/helpers/quicksight/template.py:9]: suggestion: 

```
    def __init__(self, version):
        if isinstance(version, __class__):
            self.major, self.minor, self.build = version.major, version.minor, version.build 
        elif isinstance(version, str):
            self.major, self.minor, self.build = self._parse(str_version)
        else:
            raise TypeError(f'{version} must be {__class__} or string')
```

then all other functions can be simplier

```
 def compatible_versions(self, version) -> bool:
        """
            Return True when both version are on the same major branch
        """
        return CidVersion(version).major == self.major
```

**Review by iakov-aws — CHANGES_REQUESTED** [review]
thanks for your updates. please can you remove click.echo changes and we are ready to merge

File: cid/helpers/quicksight/dashboard.py [hunk 5]
```diff
@@ -101,22 +101,38 @@ def status(self) -> str:
...
+        try:
+            cid_version_latest = self.sourceTemplate.cid_version if isinstance(self.sourceTemplate, CidQsTemplate) else "UNKNOWN"
+        except ValueError:
+            cid_version_latest = "UNKNOWN"
         if self.latest:
             print(f"  Version: {self.deployed_version}")
         else:
             print(f"  Version (deployed, latest): {self.deployed_version}, {self.latest_version}")
+            
+        print(f"  CID Version (deployed, latest): {cid_version}, {cid_version_latest}")
```
Comment (iakov-aws) [inline cid/helpers/quicksight/dashboard.py:128]: not clear what is CID version here. probably can be rephrased:

print(f" Deployed Version={cid_version}->Latest TemplateVersion= {cid_version_latest}")

File: cid/helpers/quicksight/template.py [hunk 6]
```diff
@@ -1,17 +1,72 @@
...
 
+    def __gt__(self, _version):
+        return self._get_version_as_tuple() > CidVersion(_version)._get_version_as_tuple()
+
+    def __ne__(self, _version):
+        return self._get_version_as_tuple() != CidVersion(_version)._get_version_as_tuple()
+
+    def _get_version_as_tuple(self) -> tuple:
+        return (self.major,self.minor,self.build)
 class Template(CidQsResource):
```
Comment (iakov-aws) [inline cid/helpers/quicksight/template.py:64]: Need a couple of spaces here

**Review by iakov-aws — APPROVED** [review]

**Pull request merged**